### PR TITLE
fix(utxo-lib): check prevout script against computed for non-taproot inputs

### PR DESCRIPTION
**Pull request by OttoAllmendinger**
Make sure that we are actually signing the right script.

Taproot cases will follow.

Issue: BTC-428

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/signature.ts [hunk 1]
```diff
@@ -102,6 +103,21 @@ export function getSignatureVerifications<TNumber extends number | bigint>(
     }
   }
 
+  if (
+    parsedScript.scriptType !== 'taprootKeyPathSpend' &&
+    parsedScript.scriptType !== 'taprootScriptPathSpend' &&
```
Comment (rushilbg) [inline modules/utxo-lib/src/bitgo/signature.ts:108]: Don't we have an array of taproot narrowed types already? Let's add it if not.

```
if (!isTaprootScript(parsedScript))

File: modules/utxo-lib/src/bitgo/signature.ts [hunk 2]
```diff
@@ -102,6 +103,21 @@ export function getSignatureVerifications<TNumber extends number | bigint>(
...
+    prevOutputs
+  ) {
+    const prevoutScript = prevOutputs[inputIndex].script;
```
Comment (rushilbg) [inline modules/utxo-lib/src/bitgo/signature.ts:111]:
```suggestion
    const prevOutScript = prevOutputs[inputIndex].script;
```

File: modules/utxo-lib/src/bitgo/signature.ts [hunk 3]
```diff
@@ -102,6 +103,21 @@ export function getSignatureVerifications<TNumber extends number | bigint>(
...
+
+    const output = getOutputScript(parsedScript.scriptType, parsedScript.pubScript);
+    if (!prevoutScript.equals(output)) {
+      throw new Error(
+        `prevout script ${prevoutScript.toString('hex')} does not match computed script ${output.toString('hex')}`
+      );
```
Comment (rushilbg) [inline modules/utxo-lib/src/bitgo/signature.ts:117]: We're gonna be computing the script every time we verify the existing sigs?
Although most likely a non-issue, but off the top of your head, how much overhead is this gonna add in the regular signing flow?

Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/signature.ts:117]: I think it is very small: this sort of check happens in the psbt flow by default without significant impact

File: modules/utxo-lib/src/bitgo/signature.ts (hunk 1 above)
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/signature.ts:108]: unfortunately that won't narrow down the type of `parsedScript` enough


```
export function isParsedScriptTypeTaproot(
  t: ParsedScriptType2Of3
): t is 'taprootKeyPathSpend' | 'taprootScriptPathSpend' {
  return t === 'taprootKeyPathSpend' || t === 'taprootScriptPathSpend';
}
```

image

**Review by rushilbg — APPROVED** [review]

**Pull request merged**